marlowe-chain-sync/chain-indexer/Language/Marlowe/Runtime/ChainIndexer/NodeClient.hs

```diff
@@ -46,46 +46,49 @@ import Control.Arrow ((&&&))
 import Control.Concurrent.Component
 import Control.Concurrent.STM (STM, TVar, modifyTVar, newTVar, readTVar, writeTVar)
 import Control.Monad (guard)
-import Control.Monad.Event.Class (MonadInjectEvent, emitImmediateEventArgs_, withEvent)
+import Control.Monad.Event.Class
 import Data.IntMap.Lazy (IntMap)
 import qualified Data.IntMap.Lazy as IntMap
 import Data.List (sortOn)
 import Data.Maybe (mapMaybe)
 import Data.Ord (Down(..))
 import Language.Marlowe.Runtime.ChainIndexer.Database (CardanoBlock, GetIntersectionPoints(..))
 import Observe.Event (NewEventArgs(..), addField, reference)
+import Observe.Event.Backend (setAncestorEventBackend)
 import Ouroboros.Network.Point (WithOrigin(..))
 import UnliftIO (MonadIO, MonadUnliftIO, atomically, withRunInIO)
 
 type NumberedCardanoBlock = (BlockNo, CardanoBlock)
 type NumberedChainTip = (WithOrigin BlockNo, ChainTip)
 
 -- | Describes a batch of chain data changes to write.
-data Changes = Changes
+data Changes r = Changes
   { changesRollback   :: !(Maybe ChainPoint) -- ^ Point to rollback to before writing any blocks.
   , changesBlocks     :: ![CardanoBlock]     -- ^ New blocks to write.
   , changesTip        :: !ChainTip           -- ^ Most recently observed tip of the local node.
   , changesLocalTip   :: !ChainTip           -- ^ Chain tip the changes will advance the local state to.
   , changesBlockCount :: !Int                -- ^ Number of blocks in the change set.
   , changesTxCount    :: !Int                -- ^ Number of transactions in the change set.
+  , changesEvents     :: ![r]                -- ^ References to the events that have touched these changes
```

Comment:
It seems awkward for an ADT like `Changes` to reference events, even though that isn't tied to a particular type. Cosmetically, maybe there is a generalization of this where `r` is just an annotation, not something that needs to be in a list called `changesEvents`. I don't think we need to change the existing code, however.

Reply:
I would agree if this was part of a public API, but since this is an implementation detail I think it's fine. The event reference is already a part of the API for the component anyway, so hiding or abstracting it here doesn't really gain us anything.

I think it we really wanted to hide this, we'd want to make an abstraction at the level of the concurrency framework, where all published messages from a component got tagged behind the scenes with an event reference. Something like this would be a natural fit for CSP-style concurrency with mailboxes. Then every `send` would be an event of kind `Producer` and every `recv` would be an event of kind `Consumer`.